### ISSUE TITLE
Analyzer is confused when overloaded methods change return types

### ISSUE THREAD

**Issue by stevegutz** [revapi/revapi#123]
The analyzer seems to be confused when one variant of an overloaded method has its return type changed. Here you can see that it's incorrectly reporting `java.method.numberOfParametersChanged`.

#### old
```java
public class TestOverloadedMethodChanges {

  public static void myMethod(String s, int n) {}

  public static void myMethod(int n) {}
}
```

#### new
```java
public class TestOverloadedMethodChanges {

  public static void myMethod(String s, int n) {}

  public static String myMethod(int n) {
    return "";
  }
}
```

#### results
```
Analysis results
----------------

Old API: JavaUtils-10.0-SNAPSHOT-old.jar
New API: JavaUtils-10.0-SNAPSHOT.jar
old: method void TestOverloadedMethodChanges::myMethod(int)
new: method void TestOverloadedMethodChanges::myMethod(java.lang.String, int)
java.method.numberOfParametersChanged: The number of parameters of the method have changed.
SOURCE: BREAKING, BINARY: BREAKING

old: method void TestOverloadedMethodChanges::myMethod(java.lang.String, int)
new: method java.lang.String TestOverloadedMethodChanges::myMethod(int)
java.method.numberOfParametersChanged: The number of parameters of the method have changed.
SOURCE: BREAKING, BINARY: BREAKING
java.method.returnTypeChanged: The return type changed from 'void' to  'java.lang.String'.
SOURCE: POTENTIALLY_BREAKING, BINARY: BREAKING
```

**Comment by metlos** [comment]
When resolving changes in the overloaded methods Revapi has to use some kind of heuristics to measure "similarity" of the methods in old and new so that it can match them up somehow.

In this case, yes, it looks like the heuristics could be improved :wink:

If you're interested, take a look at https://github.com/revapi/revapi/blob/master/revapi-java/src/main/java/org/revapi/java/JavaApiAnalyzer.java#L94 (which is where 2 lists of sibling elements in the AST (one from the old API and the other from the new API) get reordered according to the "correspondence" between the old and new versions). There are some tests for the overload resolution here: https://github.com/revapi/revapi/blob/master/revapi-java/src/test/java/org/revapi/java/MethodChecksTest.java#L196.

I'd love to discuss ideas (PRs even :wink:) how to improve that because frankly it's quite complex and doesn't yield great results (as you found out yourself).

**Comment by stevegutz** [comment]
Thanks for replying so quickly, Lukas. It's a tough problem for sure. One idea I had was to have a way of not allowing methods that still exist with the same signature to be used as the `old` method in certain types of checks. That would at least solve this problem, but it might not solve similar ones.

**Comment by metlos** [comment]
Should be fixed in `revapi-java:0.19.1`.

**Comment by stevegutz** [comment]
Sweet, thanks!
